test/optim/test_zeroptimizer.py

```diff
@@ -0,0 +1,469 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+import os
+
+import numpy as np
+import unittest
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+from typing import List, Any
+import io
+from torch.optim import ZeROptimizer, SGD
+skip_if_no_cuda = unittest.skipIf(not torch.cuda.is_available(), reason="cuda required")
+
+BACKEND = dist.Backend.NCCL if torch.cuda.is_available() else dist.Backend.GLOO  # type: ignore
+DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+def _broadcast_object(
+    obj: Any, src_rank: int, group: object = dist.group.WORLD, dist_device: torch.device = torch.device("cpu")
+) -> Any:
+    """
+    Either broadcast from master to the fleet (default),
+    or use the src setting as the original rank.
+    """
+
+    if dist.get_rank() == src_rank:
+        # Emit data
+        buffer = io.BytesIO()
+        torch.save(obj, buffer)
+        data = bytearray(buffer.getbuffer())
+        length_tensor = torch.LongTensor([len(data)]).to(dist_device)
+        data_send_tensor = torch.ByteTensor(data).to(dist_device)
+        dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
+        dist.broadcast(data_send_tensor, src=src_rank, group=group, async_op=False)
+    else:
+        # Fetch from the source
+        length_tensor = torch.LongTensor([0]).to(dist_device)
+        dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
+        data_recv_tensor = torch.empty([int(length_tensor.item())], dtype=torch.uint8, device=dist_device)
+        dist.broadcast(data_recv_tensor, src=src_rank, group=group, async_op=False)
+        buffer = io.BytesIO(data_recv_tensor.cpu().numpy())
+        obj = torch.load(buffer, map_location=dist_device)
+    return obj
+
+
+class TestZeROptimizer(unittest.TestCase):
+    def tearDown(self):
+        try:
+            torch.distributed.destroy_process_group()
+        except RuntimeError:
+            pass
+
+    @staticmethod
+    def dist_init(rank, world_size):
+        os.environ["MASTER_ADDR"] = "localhost"
+        os.environ["MASTER_PORT"] = "29501"
+        dist.init_process_group(backend=BACKEND, rank=rank, world_size=world_size)
+
+    def test_create(self):
+        TestZeROptimizer.dist_init(0, 1)
+        params = [torch.rand(1)]
+        o = ZeROptimizer(params, optim=SGD, lr=0.01)
+
+    def test_state_dict(self):
+        TestZeROptimizer.dist_init(0, 1)
+        x = torch.tensor([1.0], device=DEVICE, requires_grad=True)
+        o = ZeROptimizer([x], optim=SGD, lr=0.1, momentum=0.9)
+        x.backward()
+        o.step()
+        assert x == torch.tensor([0.9], device=DEVICE)
```

Comment:
Other tests use `self.assert*` to do checks. Some examples are linked below:

https://github.com/pytorch/pytorch/blob/b5662ba0f0cb6d99805a44dd5b6bcda128468c83/torch/testing/_internal/common_distributed.py#L437-L443

https://github.com/pytorch/pytorch/blob/b5662ba0f0cb6d99805a44dd5b6bcda128468c83/torch/testing/_internal/distributed/rpc/rpc_test.py#L518-L519